### ISSUE TITLE
CLOSE_WAIT connection status after disconnect clients

### ISSUE THREAD

**Issue by cdgraff** [cfreeman/signalbox#5]
Many connection still open with CLOSE_WAIT status after client disconnect

[root@SUPV6743 ~]# netstat -nap | grep :3000 | awk '{print $6}' | sort -n | uniq -c
  14698 CLOSE_WAIT
        73 ESTABLISHED
          1 LISTEN

Thanks!

**Comment by cfreeman** [comment]
Had a dig around in this one today. Found that primus ping messages were not being correctly returned, and websockets not correctly closed by the server. These two things would have quickly blasted out many open connections. 

This may also improve performance somewhat (see #4).

**Comment by aferrari-technisys** [comment]
The connections works well now! Don't see more CLOSE WAIT connections.
thanks!

**Comment by cdgraff** [comment]
Hi!

I still saw the issue:

   2567 CLOSE_WAIT
    150 ESTABLISHED
      1 LISTEN

BTW; I has stats about GO perf, numbers of GOROUTINES and more... I has captured using
https://github.com/yvasiyarov/gorelic

I see that GC take more time every minutes that application is running, the same in numbers of Go Routines and others.

I can share with you.

Let me know... thanks in advance!

**Comment by gonzafirewall** [comment]
Hello, i tested and i see that the websocket reconnect every 1.2 min only see one annunce and one ping message and close
![websocket](https://cloud.githubusercontent.com/assets/1536784/8537335/e648e0de-2433-11e5-8545-b0faf33b37f2.png)

**Comment by cdgraff** [comment]
yes I see the same issue... and if you check into webrtc of chrome see many connections chrome://webrtc-internals/

Now almost I see this message, that be new:
2015/07/07 04:17:19 ERROR - http.HandleFunc: websocket: version != 13

If I open the conections now to
http://xx.xx.37.22:3000/rtc.io/primus.js

The response is perfect but If I try to open Websocket fail.. in base to the error into the Gorilla lib, has some issues with the websocket upgrade

https://github.com/gorilla/websocket/blob/master/server.go#L97

hope this help in something

**Comment by cdgraff** [comment]
Some stats from GO process... take a look into File Descriptors used after some time...

![fd-go](https://cloud.githubusercontent.com/assets/1087447/8538423/56ad08ec-2442-11e5-9d20-6147456ef844.png)
![gc-signalbox](https://cloud.githubusercontent.com/assets/1087447/8538422/56ad0572-2442-11e5-9f85-88681903c7c5.png)

**Comment by cfreeman** [comment]
Hrmm. Thanks for the extra details. Will have another tinker around.

**Comment by cfreeman** [comment]
Are you able to send me what the agent string is in the /announce message? It is just truncated in @gonzafirewall screenshot. I just want to make sure I am testing with the same version as you.

In terms of the Gorilla lib it looks like your websocket client is not sending Sec-WebSocket-Version: 13 Are you able to send me a dump of the full client request? What is the client that is trying to connect? The browser versions that implement Websocket V13 is highlighted below:

![websockets](https://cloud.githubusercontent.com/assets/39248/8592215/e51cb002-2670-11e5-808f-5ab90f11015e.jpg)

I.e. Gorilla lib (and Signalbox) currently only support Websocket version 13.

**Comment by cdgraff** [comment]
Hi, here the logs from my Desktop.

/announce|{"id":"cde7244f-cdb4-4c8e-a36d-00403cb18c18"}|{"browser":"chrome","browserVersion":"43.0.2357","id":"cde7244f-cdb4-4c8e-a36d-00403cb18c18","agent":"signaller@2.5.1","room":"aHR0cDovL2hscy5jYXN0dG8ubWUvbGl2ZS9oVkZwVTlDZnM0VmNBNDRHb0xHRi9wbGF5bGlzdC5tM3U4"}

![pb-7cantqxpl6ymqqlbqfugiqcbdokpjxragfwrj4dk](https://cloud.githubusercontent.com/assets/1087447/8609182/b0d3a9d8-2678-11e5-8019-468fdd33f6ec.png)

I think the issue with WebSocket version, is why when we use into production with real users, some can be using older version of the browsers and IE, I think we can dismiss this message.

The main issue to check now, is the perf, the amount of FD opened... and the GC time, that always increase... after some minutes stop to response correctly the websocket part, something extrange is the amount of time that the same ID be into the announce. But at the moment of the capture, was only me using 2 tabs to simulate the p2p connection.

<img width="1280" alt="captura de pantalla 2015-07-09 20 33 12" src="https://cloud.githubusercontent.com/assets/1087447/8609307/f02958de-2679-11e5-98f5-4ccc3835a6cc.png">

Something more info that I can provide to help? just let me know?

THhnks in advance!

**Comment by cfreeman** [comment]
It looks like the application built upon rtc.io is sending additional messages on top of announce. In that snippet you sent. There is one announce and several custom /to messages -- I imagine something is trying several paths to broker a connection?

In terms of FD - from what I can tell one FD is opened per websocket. I also create one goroutine per websocket. So that would go some ways to explain the growth in FD's and Goroutines. Does the application ever send any leave commands? Is there any reason why the application would open a websocket and never close it once webrtc peers are established?

**Comment by cfreeman** [comment]
P.S. the maximum number of file descriptors that can be opened in linux can be tuned somewhat -- http://www.cyberciti.biz/faq/linux-increase-the-maximum-number-of-open-files/
